jobs/switchboard-leader/templates/readiness-switchboard.sh

```diff
@@ -0,0 +1,117 @@
+#!/bin/sh
+# Readiness probe script for the mysql-proxy role
+
+K=/var/vcap/packages/kubectl/bin/kubectl
+N="$(cat /var/run/secrets/kubernetes.io/serviceaccount/namespace)"
+H="${HOSTNAME}"
+S="<%= p('gp.switchboard.service') %>"
+D="<%= p('gp.switchboard.renewal') %>"
+A=skiff-leader
+L=/tmp/log-ready-switchboard
+
+# Clear log.
+rm $L
+
+log () { echo "$@" >> $L ; }
+
+now () {
+    # timestamp, seconds of the epoch.
+    date '+%s'
+}
+
+get_claim () {
+    log retrieval ${N}:${S} :: $A
+    export CLAIM=$($K get service -n $N $S -o "jsonpath={.metadata.annotations.$A}")
+    # CLAIM='claimant:claimtime' -- claimtime unit is [epoch].
+    log claim: $CLAIM
+    test -n "${CLAIM}"
+}
+
+our_claim () {
+    CLAIMANT=$(echo $CLAIM | awk -F: '{ print $1 }')
+    log self? $H :: $CLAIMANT
+    test "${CLAIMANT}" == "${H}"
+}
+
+is_expired () {
+    CLAIMSEC=$(echo $CLAIM | awk -F: '{ print $2 }')
+    NOW="$(now)"
+    log expired? $NOW :: $CLAIMSEC
+    test -z "${CLAIMSEC}" || test "$(( ${CLAIMSEC} + $D ))" -lt "$NOW"
+}
+
+make_claim () {
+    C=${H}:$(now)
+    log make first claim $C
+    $K annotate service -n $N $S $A=$C >> $L 2>&1
+}
+
+extend_claim () {
+    C=${H}:$(now)
+    log extend claim $C
+    $K annotate --overwrite service -n $N $S $A=$C >> $L 2>&1
+}
+
+clear_claims() {
+    log clear expired claim
+    $K annotate service -n $N $S ${A}- >> $L 2>&1
+}
+
+claim () {
+    if ! get_claim ; then
+	log no claims
+	make_claim || return 1
```

Comment:
There's an obvious race here: the claim could have been made after we failed to get it.

We might need to use `kubectl annotate --resource-version` and a bunch of resource version tracking to work around that, but that's definitely more complex code.

Reply:
In that case the `make_claim` will fail, because it does not use `--overwrite`.

Reply:
I think you're right; the races I can think of are covered given that.